src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -1420,6 +1422,196 @@ class ngraph::pass::mask_propagation::Transpose : public MatcherPass {
     }
 };
 
+class ngraph::pass::mask_propagation::VariadicSplit : public MatcherPass {
+public:
+    VariadicSplit() {
+        auto input_pattern = pattern::any_input(pattern::has_static_rank());
+        auto axis_pattern = pattern::wrap_type<opset6::Constant>();
+        auto split_lengths_pattern = pattern::wrap_type<opset6::Constant>();
+        auto split_pattern =
+            pattern::wrap_type<opset6::VariadicSplit>({input_pattern, axis_pattern, split_lengths_pattern});
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto& pattern_map = m.get_pattern_value_map();
+            auto axis_node = ngraph::as_type<opset6::Constant>(pattern_map.at(axis_pattern).get_node());
+            if (!axis_node)
+                return false;
+
+            const auto& input = pattern_map.at(input_pattern);
+            const auto input_mask = getMask(input);
+            if (!input_mask)
+                return false;
+
+            auto axis = axis_node->cast_vector<int64_t>()[0];
+            if (axis < 0)
+                axis += input_mask->size();
+
+            auto split = pattern_map.at(split_pattern).get_node();
+
+            auto split_lengths_const =
+                ngraph::as_type<opset6::Constant>(pattern_map.at(split_lengths_pattern).get_node());
+            if (!split_lengths_const)
+                return false;
+            auto split_lengths = split_lengths_const->cast_vector<int64_t>();
+
+            // adjust split_lengths if needed
+            // split_lengths can contain -1 value
+            int minus_one_length_idx = -1;
+            int64_t total_lengths = 0;
+            for (int i = 0; i < split_lengths.size(); i++) {
+                if (split_lengths[i] == -1) {
+                    minus_one_length_idx = i;
+                    continue;
+                }
+                total_lengths += split_lengths[i];
+            }
+            if (minus_one_length_idx >= 0 && (*input_mask)[axis].size() > 0) {
+                const auto& input_shape = input.get_partial_shape();
+                if (input_shape[axis].is_dynamic())
+                    return false;
+                auto split_dim = static_cast<uint64_t>(input_shape[axis].get_length());
+                split_lengths[minus_one_length_idx] = split_dim - total_lengths;
+            }
+
+            uint64_t split_start = 0;
+            uint64_t split_end = split_lengths[0];
+            std::vector<Mask::Ptr> output_masks;
+            std::vector<Mask*> output_masks_raw;
+            Mask* input_mask_raw = input_mask.get();
+            for (size_t i = 0; i < split->get_output_size(); i++) {
+                auto output_mask = std::make_shared<Mask>();
+                output_mask->copy_and_slice_mask_from(input_mask_raw, axis, split_start, split_end);
```

Comment:
Why you are modifying mask outside of callback? It's unexpected by other operations and could lead to errors during propagation

Reply:
The logic of mask propagation is following: you are initializing empty output masks, output masks are being connected with input masks by callbacks, callbacks then called at the end of the matcher to update output masks and all its neighbors. Masks is being updated only in callbacks

Reply:
This is major comment, it's quite important